modules/standard/Maps.chpl

```diff
@@ -0,0 +1,291 @@
+
+      :yields: A reference to one of the keys contained in this map.
+    */
+    iter these() const ref {
+
+    }
+
+    /*
+      Iterates over the key-value pairs of this map.
+
+      :yields: A tuple of references to one of the key-value pairs contained in
+               this map.
+    */
+    iter items() const ref {
+
+    }
+
+    /*
+      Iterates over the values of this map.
+
+      :yields: A reference to one of the values contained in this map.
+    */
+    iter values() ref {
+
+    }
+
+    /*
+      Writes the contents of this map to a channel. The format looks like:
+
+        .. code-block:: chapel
+    
+           {k1: v1, k2: v2, .... , kn: vn}
+
+      :arg ch: A channel to write to.
+    */
+    proc writeThis(ch: channel) {
+
+    }
+
+    /*
+      Adds a key-value pair to the map. Different than doing `myMap[key] =
+      value`, this method returns `false` if the key already exists in the map.
+
+     :arg k: The key to add to the map
+     :type k: keyType
+
+     :arg v: The value that maps to ``k``
+     :type k: valueType
+
+     :returns: `true` if `k` was not in the map and added with value `v`.
+               `false` otherwise.
+     :rtype: bool
+    */
+    proc add(k: keyType, v: valType): bool {
+
+    }
+
+    /*
+      Removes a key-value pair to the map, withe the given key.
+      
+     :arg k: The key to remove from the map
+     :type k: keyType
+
+     :returns: `false` if `k` was not in the map.  `true` if it was and removed.
+     :rtype: bool
+    */
+    proc remove(k: keyType): bool {
+
+    }
+
+    /*
+      Returns a new 0-based array containing a copy of key-value pairs as
+      tuples.
+
+      :return: A new DefaultRectangular array.
+      :rtype: [] (keyType, valType)
+    */
+    proc toArray(): [] (keyType, valType) {
+
+    }
+
+    /*
+      Returns a new 0-based array containing a copy of keys.
+
+      :return: A new DefaultRectangular array.
+      :rtype: [] keyType
+    */
+    proc keysToArray(): [] keyType {
+
+    }
+
+    /*
+      Returns a new 0-based array containing a copy of values.
+
+      :return: A new DefaultRectangular array.
+      :rtype: [] valType
+    */
+    proc valuesToArray(): [] valType {
+
+    }
+
+  } // end record map
+
+  /*
+    Replace the content of this map with the other's.
+
+    .. warning::
+
+      This will invalidate any references to elements previously contained in
+      `lhs`.
+
+    :arg lhs: The map to assign to.
+    :arg rhs: The map to assign from. 
+  */
+  proc =(ref lhs: map(?kt, ?vt), const ref rhs: map(kt, vt)){
+
+  }
+
+
+  /*
+    Returns `true` if the contents of two maps are the same.
+
+    :arg a: A map to compare.
+    :type a: map
+
+    :arg b: A map to compare.
+    :type b: map (with same keyType and valType)
+
+    :return: `true` if the contents of two maps are equal.
+    :rtype: `bool`
+  */
+  proc ==(const ref a: map(?k, ?v), const ref b: map(?k, ?v)): bool {
```

Comment:
Are we comparing on the basis of keys or values? Both?

Reply:
The reason I brought this up, was because in most set implementations I've seen, equality `==` is usually only required over the _key_ type (along with a hash function of some sort). In a world where we had interfaces, we might say that the key type `K` must be `Comparable` and `Hashable`, but we obviously don't have that yet.

Reply:
I don't think this is true for Python or C++. i.e. in both of those map equality means key-value mappings are identical, too. Am I missing something? Or what are the languages where map equality means equality of key sets?

Reply:
I would expect, even with interfaces, that we could have `==` compare keys and values and just insist at this point that the value type has a `==` available. The map would still work fine if the value type didn't have `==` available; you just wouldn't be able to call this method.

Reply:
I'm fine with just assuming both key type and value type have a `==` defined.